File: 110-balanced-binary-tree/110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxDiff = 0;
    public boolean isBalanced(TreeNode root) {
        if(root == null) return true;
        else {
            findHeight(root, 0);
            return maxDiff <= 1;
        }
    }
    
    private int findHeight(TreeNode root, int h) {
        if(root == null) return h;
        
        int leftH = findHeight(root.left, h);
        int rightH = findHeight(root.right, h);
        
        maxDiff = Math.max(maxDiff, Math.abs(leftH - rightH));
        return Math.max(leftH, rightH) + 1;
    }
}

File: 110-balanced-binary-tree/110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return check(root) == -1 ? false : true;
    }
    
    int check(TreeNode *root) {
        if(!root) return 0;
        
        int left = check(root -> left);
        if(left == -1) return -1;
        
        int right = check(root -> right);
        if(right == -1) return -1;
        
        if(abs(left - right) > 1) return -1;
        else return max(left, right) + 1;
    }
};